Earn: Convert ads form to typescript

edanzer opened the pull request:
## Proposed Changes

Refactors Earn > Ads > FormSettings component to TypeScript.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1) Checkout this branch and go to `http://calypso.localhost:3000/earn/YOURDOMAIN`
2) If you haven't yet, find the Ads card, enable Ads, and on the Ads page click the button to apply (it is instant).
3) Go to Ads > Settings. Confirm the page loads without issues. 
4) Confirm initial settings (if you haven't already changed them) should look like this: 
<img width="1056" alt="ad-settings" src="https://github.com/Automattic/wp-calypso/assets/21228350/aaa046d9-16ba-4502-a40b-baba46b45455">
5) Try updating various settings and saving. Confirm your new options are saved. 

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

edanzer committed:
Convert Earn Ads form to typescript

edanzer committed:
Add comment for undefined classnames

simison commented on client/my-sites/earn/ads/form-settings.tsx:
```diff
@@ -176,6 +206,8 @@ const AdsFormSettings = () => {
 						disabled={ isLoading }
 						label={ translate( 'Run ads only for logged-out users (less revenue)' ) }
+						// className undefined to resolve TS warning
+						className={ undefined }
```
huh? 🤔 `FormRadio` requires `className` but it can be `undefined` — why? Could it be made optional at component level instead?

edanzer replied:
@simison - One other way to handle this is just to add TS comment exceptions as being done [here](https://github.com/Automattic/wp-calypso/blob/9f30cbb0c8f3e91664bcac14b9b25177b498ba61/client/my-sites/domains/domain-management/settings/cards/domain-forwarding-card.tsx#L476) and [here](https://github.com/Automattic/wp-calypso/blob/9f30cbb0c8f3e91664bcac14b9b25177b498ba61/client/my-sites/site-settings/reading-newsletter-settings/ExcerptSetting.tsx#L33). 

We could (and probably should) refactor some of these underlying components to not require className, and I considered doing so here.

BUT... it would dramatically expand the reach and risk of this PR if it included refactoring underlying, common Automattic components, which are themselves used throughout the code base. My goal in each of these refactors is to keep the size/risk small. If we felt it was worth updating the underlying components, I'd rather do that in separate PRs (one per component that needs refactoring). 

See other [comment](https://github.com/Automattic/wp-calypso/pull/81609#issuecomment-1716236323) for more context on the issue.

edanzer replied:
Just to follow convention with other uses of this component, I've updated the PR to just comment out the TS errors for now.

simison reviewed (approved):
Tests well and looks good otherwise but I don't understand why we needed `className` addition.

Feel free to merge.

edanzer commented:
> I don't understand why we needed `className` addition.

Just to be sure about the underlying issue. We've updated this file to TS, but the underlying shared components are in JS. When this TS code here sees something like this (from FormRadio component [here](https://github.com/Automattic/wp-calypso/blob/9f30cbb0c8f3e91664bcac14b9b25177b498ba61/client/components/forms/form-radio/index.jsx#L5)

`const FormRadio = ( { className, label, ...otherProps } ) => () `

...it interprets that to mean that className and label are required. If we updated the underlying component to TS, we'd specify those props are optional. But that requires refactoring our underlying shared components. 

I should note I initially just tried passing `className=''` which is less odd looking than undefined, but this throws a calypso-specific ESLint error that className needs to abide by BEM syntax. I tried commenting that out, but that also got messy. So short of other messy comments or refactoring the underlying components (high risk), this was the least bad solution I could find.

edanzer committed:
Add TS exception comments
